compiler/src/dmd/cparse.d

```diff
@@ -3159,6 +3174,83 @@ final class CParser(AST) : Parser!AST
         return cast(AST.StringExp) label;
     }
 
+    /********************
+     * Parse C inline assembler statement.
+     * gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html
+     * Returns:
+     *   inline assembler expression as a Statement
+     */
+    private AST.Statement cparseAsm()
+    {
+        // Defer parsing of AsmStatements until semantic processing.
+        const loc = token.loc;
+
+        nextToken();
+
+        // Consume all asm-qualifiers. As a future optimization, we could record
+        // the `inline` and `volatile` storage classes against the statement.
+        while (token.value == TOK.goto_ ||
+               token.value == TOK.inline ||
+               token.value == TOK.volatile)
+            nextToken();
+
+        check(TOK.leftParenthesis);
+        Token* toklist = null;
+        Token** ptoklist = &toklist;
+        Identifier label = null;
+        auto statements = new AST.Statements();
+        while (1)
+        {
+            switch (token.value)
+            {
+                case TOK.rightParenthesis:
+                    break;
+
+                case TOK.semicolon:
+                    error("matching `)` expected, not `;`");
+                    break;
+
+                case TOK.endOfFile:
+                    /* ( */
+                    error("matching `)` expected, not end of file");
+                    break;
+
+                case TOK.colonColon:  // treat as two separate : tokens for iasmgcc
```

Comment:
Instead of trying to unwind ::, I suggest having the lexer not create a TOK.colonColon if compiling a C file. Trying to "unwind" the TOK.colonColon is stepping on the turf of Lexer.